### ISSUE TITLE
Parquet reader "children size should not be larger than 2" error

### ISSUE THREAD

**Issue by chliang71** [facebookincubator/velox#7002]
### Bug description

We encounter an issue where there Parquet reader was getting "children size should not be larger than 2" error on a nested column of array<struct> type. The Parquet file is a valid (non-corrupted) file, readable from Java implmentation and parquet-tools so 

The column that hit this issue has the definition like the following:
level1_column_X - (type struct)
---- |-------- level2_column - array<struct>
----------------------|----------- "array"
-------------------------------------|----------- level3_column1 - string
-------------------------------------|----------- level3_column2 - string
-------------------------------------|----------- level3_column3 - string
-------------------------------------|----------- level3_column4 - string
-------------------------------------|----------- level3_column5 - string

The problem happened when reading the "array" part, as it has 5 children, while the code [here](https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/reader/ParquetReader.cpp) expects no more than two. Attaching the log from GDB below. It is showing the "array" part schemaElement did have 
repetition_type = true,
converted_type = false,
repetition_type = facebook::velox::parquet::thrift::FieldRepetitionType::REPEATED
num_children = 5

Based on the code, it looks to me the gap here is that the Parquet reader seems to be expecting something like below instead:

level1_column_X - (type struct)
---- |-------- level2_column - array<struct>
----------------------|----------- "array"
--------------------------------------|----------- "bag"
-----------------------------------------------------|----------- level3_column1 - string
-----------------------------------------------------|----------- level3_column2 - string
-----------------------------------------------------|----------- level3_column3 - string
-----------------------------------------------------|----------- level3_column4 - string
-----------------------------------------------------|----------- level3_column5 - string

Basically, one more node between "array" and its children.

I tried to create a Parquet file myself from Presto with a similar definition to try to reproduce this issue, but the file I created does not have this issue, as the file did have the "bag" part. I wonder if this related to Parquet etc.

### System information

Velox System Info v0.0.2
Commit: f042f8b4e1c86fa1729f819a1f505965398c4d61
CMake Version: 3.27.0
System: Linux-5.10.176+
Arch: x86_64
C++ Compiler: /usr/bin/c++
C++ Compiler Version: 12.2.0
C Compiler: /usr/bin/cc
C Compiler Version: 12.2.0
CMake Prefix Path: /usr/local;/usr;/;/usr/local;/usr/local;/usr/X11R6;/usr/pkg;

### Relevant logs

```bash
(gdb) p schemaElement
$4 = (facebook::velox::parquet::thrift::SchemaElement &) @0x26022f0: {<apache::thrift::TBase> = {_vptr.TBase = 0x2505dd8 <vtable for facebook::velox::parquet::thrift::SchemaElement+48>},
  type = facebook::velox::parquet::thrift::Type::BOOLEAN, type_length = 0, repetition_type = facebook::velox::parquet::thrift::FieldRepetitionType::REPEATED, name = {
    _M_dataplus = {<std::allocator<char>> = {<std::__new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x2602318 "array"}, _M_string_length = 5, {
      _M_local_buf = "array\000\000\0000#`\002\000\000\000", _M_allocated_capacity = 521325933153}}, num_children = 5, converted_type = facebook::velox::parquet::thrift::ConvertedType::UTF8, scale = 0,
  precision = 0, field_id = 0, logicalType = {<apache::thrift::TBase> = {_vptr.TBase = 0x2505e40 <vtable for facebook::velox::parquet::thrift::LogicalType+48>}, STRING = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2506528 <vtable for facebook::velox::parquet::thrift::StringType+48>}, <No data fields>}, MAP = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2506458 <vtable for facebook::velox::parquet::thrift::MapType+48>}, <No data fields>}, LIST = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x25063f0 <vtable for facebook::velox::parquet::thrift::ListType+48>}, <No data fields>}, ENUM = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2506388 <vtable for facebook::velox::parquet::thrift::EnumType+48>}, <No data fields>}, DECIMAL = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2506250 <vtable for facebook::velox::parquet::thrift::DecimalType+48>}, scale = 0, precision = 0}, DATE = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2506320 <vtable for facebook::velox::parquet::thrift::DateType+48>}, <No data fields>}, TIME = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2505fe0 <vtable for facebook::velox::parquet::thrift::TimeType+48>}, isAdjustedToUTC = false, unit = {<apache::thrift::TBase> = {
          _vptr.TBase = 0x25060b0 <vtable for facebook::velox::parquet::thrift::TimeUnit+48>}, MILLIS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x25061e8 <vtable for facebook::velox::parquet::thrift::MilliSeconds+48>}, <No data fields>}, MICROS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x2506180 <vtable for facebook::velox::parquet::thrift::MicroSeconds+48>}, <No data fields>}, NANOS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x2506118 <vtable for facebook::velox::parquet::thrift::NanoSeconds+48>}, <No data fields>}, __isset = {MILLIS = false, MICROS = false, NANOS = false}}},
    TIMESTAMP = {<apache::thrift::TBase> = {_vptr.TBase = 0x2506048 <vtable for facebook::velox::parquet::thrift::TimestampType+48>}, isAdjustedToUTC = false, unit = {<apache::thrift::TBase> = {
          _vptr.TBase = 0x25060b0 <vtable for facebook::velox::parquet::thrift::TimeUnit+48>}, MILLIS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x25061e8 <vtable for facebook::velox::parquet::thrift::MilliSeconds+48>}, <No data fields>}, MICROS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x2506180 <vtable for facebook::velox::parquet::thrift::MicroSeconds+48>}, <No data fields>}, NANOS = {<apache::thrift::TBase> = {
            _vptr.TBase = 0x2506118 <vtable for facebook::velox::parquet::thrift::NanoSeconds+48>}, <No data fields>}, __isset = {MILLIS = false, MICROS = false, NANOS = false}}},
    INTEGER = {<apache::thrift::TBase> = {_vptr.TBase = 0x2505f78 <vtable for facebook::velox::parquet::thrift::IntType+48>}, bitWidth = 0 '\000', isSigned = false}, UNKNOWN = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x25062b8 <vtable for facebook::velox::parquet::thrift::NullType+48>}, <No data fields>}, JSON = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2505f10 <vtable for facebook::velox::parquet::thrift::JsonType+48>}, <No data fields>}, BSON = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x2505ea8 <vtable for facebook::velox::parquet::thrift::BsonType+48>}, <No data fields>}, UUID = {<apache::thrift::TBase> = {
        _vptr.TBase = 0x25064c0 <vtable for facebook::velox::parquet::thrift::UUIDType+48>}, <No data fields>}, __isset = {STRING = false, MAP = false, LIST = false, ENUM = false, DECIMAL = false,
      DATE = false, TIME = false, TIMESTAMP = false, INTEGER = false, UNKNOWN = false, JSON = false, BSON = false, UUID = false}}, __isset = {type = false, type_length = false, repetition_type = true,
    num_children = true, converted_type = false, scale = false, precision = false, field_id = false, logicalType = false}}
(gdb) p name
$5 = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<std::__new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7fffffffc8c0 "array"},
  _M_string_length = 5, {_M_local_buf = "array\000\000\000\030\313\377\377\377\177\000", _M_allocated_capacity = 521325933153}}
```

**Comment by chliang71** [comment]
@yingsu00 @liushengxuan much appreciate any insights on this. thanks

**Comment by yingsu00** [comment]
@chliang71 Yes Velox does expect the "bag" child. What tool did you use to create the Parquet file with this problem? Can you create a file that can reproduce the issue?

**Comment by chliang71** [comment]
Thanks for confirming. I believe this file is ingested from Spark but I will dig into it and try to see if you have reproduce with a testing file. And just for the record, parquet-tools inspect command shows info below

############ file meta data ############
created_by: parquet-mr version 1.10.35 (build c730a48e56e7f87bedd1a8360ee8656628fc971c)
num_columns: 173
num_rows: 133814
num_row_groups: 2
format_version: 1.0
serialized_size: 943221

**Comment by liushengxuan** [comment]
Thanks @chliang71 for raising the problem and @yingsu00 for helping. This happens at the footer parsing stage. If you are not able to provide the whole Parquet file, you can extract the footer only.

**Comment by yingsu00** [comment]
@chliang71 Any update on getting the test file footer?

**Comment by qqibrow** [comment]
@yingsu00 I am able to get a test file footer. working on the fix. thanks.

**Comment by chliang71** [comment]
We did some more investigation and debug, and now we think that the fix might be that, for [here](https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/reader/ParquetReader.cpp#L311) 

instead of checking 
children.size = 1 -> LIST
children.size = 2 -> MAP

looks to us that it should be more like below:
parentSchemaElement.converted_type == LIST -> LIST
parentSchemaElement.converted_type == MAP -> MAP

Basically, looks like for the dummy "array" and "map" nodes, their parent schema element actually has the type info regarding LIST or MAP, so we don't have to make the inference based on their own children sizes. This change fixed the issue for our previously failing Parquet files. We will test more and file a PR for review later. For now, FYI @yingsu00 @qqibrow would love to hear if you have any comments, thanks!

**Comment by qqibrow** [comment]
Attach a test file can reproduce the issue:  https://www.dropbox.com/scl/fi/mo68sj3n9flnnfucmav8x/native_parquet_reader_test179106142756377866parquet?rlkey=w0vfmheq6986cpxn9tv3hbcj2&dl=0 
```
lniu@devrestricted-lniu:~/velox_parquet_test_triage/fail_parquet_files/testCustomSchemaArrayOfStructs$ parquet-tools show native_parquet_reader_test179106142756377866parquet
+------------------------------------------------------------------------+
| self                                                                   |
|------------------------------------------------------------------------|
| []                                                                     |
| [{'a': 1.0, 'b': None, 'c': '0'} {'a': None, 'b': True, 'c': '1'}]     |
| [{'a': 3.0, 'b': False, 'c': '2'} {'a': 5.0, 'b': None, 'c': '3'}      |
|  {'a': None, 'b': None, 'c': '4'}]                                     |
| [{'a': None, 'b': True, 'c': '5'}]                                     |
| [{'a': None, 'b': False, 'c': '6'}]                                    |
| [{'a': 7.0, 'b': None, 'c': '7'}]                                      |
| [{'a': 11.0, 'b': True, 'c': '8'}]                                     |
| [{'a': None, 'b': False, 'c': '9'}]                                    |
| [{'a': 13.0, 'b': None, 'c': '10'}]                                    |
| [{'a': 17.0, 'b': None, 'c': '11'}]                                    |
| [{'a': 1.0, 'b': True, 'c': '12'} {'a': None, 'b': False, 'c': '13'}]  |
| [{'a': 3.0, 'b': None, 'c': '14'}]                                     |
| [{'a': 5.0, 'b': True, 'c': '15'}]                                     |
| [{'a': None, 'b': False, 'c': '16'} {'a': None, 'b': None, 'c': '17'}] |
| [{'a': None, 'b': None, 'c': '18'} {'a': 7.0, 'b': True, 'c': '19'}]   |
| []                                                                     |
| [{'a': 11.0, 'b': False, 'c': '20'}]                                   |
| [{'a': None, 'b': None, 'c': '21'}]                                    |
| []                                                                     |
| [{'a': 13.0, 'b': True, 'c': '22'} {'a': 17.0, 'b': False, 'c': '23'}] |
| [{'a': 1.0, 'b': None, 'c': '24'} {'a': None, 'b': None, 'c': '25'}    |
|  {'a': 3.0, 'b': True, 'c': '26'} {'a': 5.0, 'b': False, 'c': '27'}    |
|  {'a': None, 'b': None, 'c': '28'}]                                    |
| []                                                                     |
+------------------------------------------------------------------------+
```

**Comment by qqibrow** [comment]
@chliang71 thanks! how do you test it? I tried to patch your diff and now it shows a different error. It might imply the inferred type is not correct: 
```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.079 s <<< FAILURE! - in com.facebook.presto.hive.parquet.TestParquetReader
[ERROR] com.facebook.presto.hive.parquet.TestParquetReader.testCustomSchemaArrayOfStructs  Time elapsed: 1.035 s  <<< FAILURE!
java.lang.UnsupportedOperationException: com.facebook.presto.common.block.LongArrayBlock
        at com.facebook.presto.common.block.Block.getBlock(Block.java:91)
        at com.facebook.presto.common.type.RowType.getObject(RowType.java:170)
        at com.facebook.presto.common.type.RowType.getObjectValue(RowType.java:146)
        at com.facebook.presto.common.type.ArrayType.arrayBlockToObjectValues(ArrayType.java:160)
        at com.facebook.presto.common.type.ArrayType.lambda$getObjectValue$0(ArrayType.java:147)
        at com.facebook.presto.common.block.AbstractArrayBlock.apply(AbstractArrayBlock.java:298)
        at com.facebook.presto.common.type.ArrayType.getObjectValue(ArrayType.java:147)
        at com.facebook.presto.hive.parquet.ParquetTester.decodeObject(ParquetTester.java:963)
        at com.facebook.presto.hive.parquet.ParquetTester.assertPageSource(ParquetTester.java:693)
        at com.facebook.presto.hive.parquet.ParquetTester.assertPageSource(ParquetTester.java:664)
        at com.facebook.presto.hive.parquet.ParquetTester.assertNativeReaderFileContents(ParquetTester.java:633)
        at com.facebook.presto.hive.parquet.ParquetTester.assertContents(ParquetTester.java:553)
        at com.facebook.presto.hive.parquet.ParquetTester.assertRoundTrip(ParquetTester.java:400)
        at com.facebook.presto.hive.parquet.ParquetTester.testRoundTripType(ParquetTester.java:346)
        at com.facebook.presto.hive.parquet.ParquetTester.testRoundTrip(ParquetTester.java:327)
        at com.facebook.presto.hive.parquet.ParquetTester.testSingleLevelArrayRoundTrip(ParquetTester.java:312)
        at com.facebook.presto.hive.parquet.AbstractTestParquetReader.testCustomSchemaArrayOfStructs(AbstractTestParquetReader.java:236)
        at com.facebook.presto.hive.parquet.TestParquetReader.testCustomSchemaArrayOfStructs(TestParquetReader.java:69)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
        at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
        at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
        at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
        at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
        at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
 ```
Could you try setup the testing infra yourself and test it? https://gist.github.com/qqibrow/689ed97b91cc0b58337be96a86291301

**Comment by qqibrow** [comment]
Overall, I feel the key here is handling Backward-compatibility rules for complex types. This is what we can refer to in the past:

https://github.com/prestodb/presto/pull/9156/commits/592b74ad3f4ea189702a4fe0b4695cecf2aaacd3

**Comment by chliang71** [comment]
Thanks for the info! sure I will take a look with this file. For test, we ran it against our internal Parquet file which was previously failing and now seemed to have passed.

**Comment by yingsu00** [comment]
This should be the backward compatibility issue. The fix needs to be together with https://github.com/facebookincubator/velox/pull/7812 for the map type, in which we need to add a parant parameter to the function.

**Comment by yingsu00** [comment]
Fixed by https://github.com/facebookincubator/velox/pull/7812. Closing.
